src/Resources/skeleton/authenticator/Security51LoginFormAuthenticator.tpl.php

```diff
@@ -0,0 +1,80 @@
+<?= "<?php\n" ?>
+
+namespace <?= $namespace ?>;
+
+use App\Repository\UserRepository;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Http\Authenticator\AbstractLoginFormAuthenticator;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\CsrfTokenBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\PasswordUpgradeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\PasswordCredentials;
+use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+use Symfony\Component\Security\Http\Util\TargetPathTrait;
+
+class <?= $class_name; ?> extends AbstractLoginFormAuthenticator
+{
+    use TargetPathTrait;
+
+    public const LOGIN_ROUTE = 'app_login';
+
+    private <?= $use_typed_properties ? 'UserRepository ' : null ?>$userRepository;
+    private <?= $use_typed_properties ? 'UrlGeneratorInterface ' : null ?>$urlGenerator;
+
+    public function __construct(UserRepository $userRepository, UrlGeneratorInterface $urlGenerator)
+    {
+        $this->userRepository = $userRepository;
+        $this->urlGenerator = $urlGenerator;
+    }
+
+    protected function getLoginUrl(Request $request): string
+    {
+        return $this->urlGenerator->generate(self::LOGIN_ROUTE);
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return self::LOGIN_ROUTE === $request->attributes->get('_route')
+            && $request->isMethod('POST');
+    }
+
+    public function authenticate(Request $request): PassportInterface
+    {
+        /** @var UserInterface|null $user */
+        $user = $this->userRepository->findOneBy(['username' => $request->get('username')]);
+
+        if (null === $user) {
+            throw new CustomUserMessageAuthenticationException('Invalid Credentials.');
+        }
+
+        /** @var string */
+        $password = $request->get('password', '');
+
+        /** @var string|null */
+        $csrf = $request->get('_csrf_token');
+
+        $credentials = new PasswordCredentials($password);
+
+        return new Passport(
+            $user, $credentials, [
```

Comment:
Since 5.2, we need to pass a UserBadge instead of User. Something like this
```
new UserBadge($credentials['username'], function ($username) {
    $user = $this->userProvider->loadUserByUsername($username);
    if (!$user instanceof UserInterface) {
        throw new AuthenticationServiceException('The user provider must return a UserInterface object.');
    }

    return $user;
})
```